### PR TITLE
Adding Doctrine examples

### PR DESCRIPTION
**Pull request by ghost**
1. `Extraload\Extractor\Doctrine\QueryExtractor` is added
2. `examples/05_default_doctrine_statement_console.php` is created
3. `README` update

### CODE REVIEW

**Comment by ghost** [comment]
- Just added a Doctrine `PreparedQueryExtractor`

File: README.md [hunk 1]
```diff
@@ -21,9 +21,9 @@
 Powerful ETL library.
 
 
-## Example
+## 1. Examples
```
Comment (umpirsky) [inline README.md]: Why numbers? :)

File: examples/06_default_doctrine_prepared_query_console.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+use Doctrine\DBAL\Connection;
+use Extraload\Extractor\Doctrine\PreparedQueryExtractor;
+use Extraload\Loader\ConsoleLoader;
+use Extraload\Pipeline\DefaultPipeline;
+use Extraload\Transformer\NoopTransformer;
+use Symfony\Component\Console\Helper\Table;
+use Symfony\Component\Console\Output\ConsoleOutput;
+
+$config = new \Doctrine\DBAL\Configuration();
+$connectionParams = array(
```
Comment (umpirsky) [inline examples/06_default_doctrine_prepared_query_console.php]: You can share `$connectionParams` between your examples, it will be easier to configure only once and be able to run all examples that way.

File: src/Extraload/Extractor/Doctrine/PreparedQueryExtractor.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            $stmt->bindValue(
+                $value['parameter'],
+                $value['value'],
+                $value['data_type'] ?? null
+            );
+        }
+
+        $stmt->execute();
+
+        $this->data = $stmt->fetchAll();
```
Comment (umpirsky) [inline src/Extraload/Extractor/Doctrine/PreparedQueryExtractor.php]: You should fetch inside `extract()`, as late as possible.

Comment (umpirsky) [inline src/Extraload/Extractor/Doctrine/PreparedQueryExtractor.php]: For large data sets, and that's common in ETL, fetching all at once can be impossible to process and lead to out of memory issues, especially if doing heavy transformations. It's better if we can extract row by row, and only keep one row in memory at a time.

File: src/Extraload/Extractor/Doctrine/QueryExtractor.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+class QueryExtractor implements ExtractorInterface
+{
+    private $position = 0;
+
+    private $data;
+
+    public function __construct(Connection $conn, string $sql)
+    {
+        $this->position = 0;
+        $this->data = $conn->query($sql)->fetchAll();
```
Comment (umpirsky) [inline src/Extraload/Extractor/Doctrine/QueryExtractor.php]: Same applies here.

File: src/Extraload/Extractor/Doctrine/QueryExtractor.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace Extraload\Extractor\Doctrine;
+
+use Doctrine\DBAL\Connection;
+use Extraload\Extractor\ExtractorInterface;
+
+class QueryExtractor implements ExtractorInterface
```
Comment (umpirsky) [inline src/Extraload/Extractor/Doctrine/QueryExtractor.php:8]: One of this two classes is redundant. I'd refactor it to only `QueryExtractor`.

`(Connection $conn, string $sql, array $values)` will do the same job if `$values = []`.

**Review by umpirsky — CHANGES_REQUESTED** [review]
Great job, I would recommand two things:

- Do one row at a time.
- Merge this two implementations into one.

Thanks!

File: README.md (hunk 1 above)
Comment (ghost) [reply README.md]: Okay no problem :)

File: examples/06_default_doctrine_prepared_query_console.php (hunk 2 above)
Comment (ghost) [reply examples/06_default_doctrine_prepared_query_console.php]: Yeah, let's do this. Thanks.

File: src/Extraload/Extractor/Doctrine/PreparedQueryExtractor.php (hunk 3 above)
Comment (ghost) [reply src/Extraload/Extractor/Doctrine/PreparedQueryExtractor.php]: Thank you very much for your feedback.

File: src/Extraload/Extractor/Doctrine/QueryExtractor.php (hunk 5 above)
Comment (ghost) [reply src/Extraload/Extractor/Doctrine/QueryExtractor.php:8]: I believe my original suggestion brings a bit of awareness to the API user about the fact that they are using either queries or prepared queries.

The logic being run under the hood is a bit different. `QueryExtractor` runs `$conn->query($sql);` whereas `PreparedQueryExtractor` runs `$conn->prepare($sql);`.

**Comment by ghost** [comment]
Thanks for the review! `QueryExtractor` and `PreparedQueryExtractor` are merged now into `AbstractExtractor` if this makes more sense.

**Comment by ghost** [comment]
Just added one more Doctrine example using a `MapFieldsTransformer` -- btw I won't push any more examples to this PR until the `MapFieldsTransformer` example is reviewed.

File: src/Extraload/Transformer/MapFieldsTransformer.php [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+class MapFieldsTransformer implements TransformerInterface
+{
+    private $fields;
+
+    public function __construct(array $fields)
+    {
+        $this->fields = $fields;
+    }
+
+    public function transform($data)
```
Comment (umpirsky) [inline src/Extraload/Transformer/MapFieldsTransformer.php]: I like this transformer, but we can make it better by using https://symfony.com/doc/current/components/property_access.html and make it work with multi-dimensional arrays and objects. What do you think?

Also, it would be great to extract transformet in a separate PR.

**Review by umpirsky — COMMENTED** [review]
Why can't we have one class doing both query with and without parameters?

**Comment by ghost** [comment]
- The `MapFieldsTransformer` is temporarily removed from this PR
- `QueryExtractor` and `PreparedQueryExtractor` are merged into `QueryExtractor` -- which is now a prepared query with 3 params with the third one being optional

**Review by umpirsky — APPROVED** [review]
My only concern is that `$this->data` can become huge. But I guess we can optimize later. 

Thanks!

**Comment by ghost** [comment]
Yeah, it makes sense. `$this->data` can always be bigger and bigger. Perhaps that is a problem that may be addressed by splitting huge queries into smaller chunks to multiprocess them in parallel.
